nri-kafka/src/Kafka/Worker.hs

```diff
@@ -0,0 +1,505 @@
+{-# LANGUAGE GADTs #-}
+
+-- The Kafka worker has the following concurrent workflows:
+-- 1. The main thread which handles
+--   - gracefully quitting
+--   - a single-thread `pollingLoop` that reads new messages from kafka
+-- 2. The Kafka `rebalanceCallback` which handles rebalancing partitions, in turn
+--    turning on and off worker threads.
+-- 3. A single-thread `pauseAndAnalyticsLoop` that tells the kafka library to pause & resume
+--    sending us messages from specific partitions (based on a number of factors)
+-- 4. Multiple worker-threads (`Partition.processMsgLoop`), one per
+--    (topic,partition)
+
+-- | Kafka.Worker is a module for processing a Kafka log.
+-- It can be used to build a CLI that will consume and process a user-defined message type
+module Kafka.Worker
+  ( process,
+    Description (..),
+    Partition.MessageCallback (Partition.MessageCallback),
+    OffsetSource (..),
+    Consumer.ConsumerGroupId (..),
+    PartitionKey,
+    Consumer.TopicName (..),
+    Consumer.ConsumerRecord (..),
+    Partition.ProcessResult (..),
+    Consumer.PartitionId (..),
+    Consumer.unOffset,
+    Settings.Settings,
+    Settings.decoder,
+    -- For tests
+    processWithoutShutdownEnsurance,
+  )
+where
+
+import qualified CLI
+import qualified Conduit
+import qualified Control.Concurrent
+import qualified Control.Concurrent.Async as Async
+import qualified Control.Concurrent.STM as STM
+import qualified Control.Concurrent.STM.TVar as TVar
+import qualified Control.Exception.Safe as Exception
+import qualified Data.UUID
+import qualified Data.UUID.V4
+import qualified Dict
+import qualified GHC.Clock
+import qualified Kafka
+import qualified Kafka.Consumer as Consumer
+import qualified Kafka.Metadata
+import qualified Kafka.Worker.Analytics as Analytics
+import qualified Kafka.Worker.Fetcher as Fetcher
+import qualified Kafka.Worker.Partition as Partition
+import qualified Kafka.Worker.Settings as Settings
+import qualified Kafka.Worker.Stopping as Stopping
+import qualified Observability
+import qualified Set
+import qualified System.Exit
+import qualified System.Posix.Process
+import qualified System.Posix.Signals as Signals
+import qualified Prelude
+
+-- | Alias for a TopicName and PartitionId, something every message will have
+type PartitionKey = (Consumer.TopicName, Consumer.PartitionId)
+
+type AllPartitions = TVar.TVar (Dict.Dict PartitionKey Partition.Partition)
+
+data Rebalance = Assign | Revoking | Revoked deriving (Show)
+
+type RebalanceInfo = TVar.TVar (Dict.Dict PartitionKey (Rebalance, Float))
+
+data State = State
+  { partitions :: AllPartitions,
+    -- | When we receive a shutdown signal this variable will change. Threads
+    -- should stop processing new messages to allow the process to shut down.
+    stopping :: Stopping.Stopping,
+    analytics :: Analytics.Analytics,
+    rebalanceInfo :: RebalanceInfo
+  }
+
+-- | All the required parameters to start the Kafka worker
+data Description = Description
```

Comment:
:art: Feels weird that we have both a `Settings` type and then a separate `Description` type. Can we combine these?